### PR TITLE
accessible buffer polish

### LINKED ISSUE #173734: Move terminal buffer element request into provider callback

### LINKED ISSUE #173780: When I'm in the terminal's accessibility buffer and I want to move focus to the terminal, I need to press the tab key twice
Type: <b>Bug</b>
CC @meganrogge 
CC @isidorn 

Using orca I open a terminal in vscode, type some commands, press shift+tab and explore the result using the arrow keys. After that I need to press the tab key twice to move the focus to the terminal. Previously I only had to press the tab key once.

VS Code version: Code - Insiders 1.76.0-insider (a7d49b0f35f50e460835a55d20a00a735d1665a3, 2023-02-08T05:24:48.579Z)
OS version: Linux x64 6.1.9-200.fc37.x86_64
Modes:
Sandboxed: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (8 x 2999)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|1, 1, 1|
|Memory (System)|15.27GB (8.27GB free)|
|Process Argv|--crash-reporter-id bc02adbd-6c3a-44b8-af92-55103a608af7|
|Screen Reader|yes|
|VM|0%|
|DESKTOP_SESSION|gnome-xorg|
|XDG_CURRENT_DESKTOP|GNOME|
|XDG_SESSION_DESKTOP|gnome-xorg|
|XDG_SESSION_TYPE|x11|
</details><details><summary>Extensions (37)</summary>

Extension|Author (truncated)|Version
---|---|---
doxdocgen|csc|1.4.0
vscode-markdownlint|Dav|0.49.0
vscode-chatgpt|gen|3.5.6
rest-client|hum|0.25.1
better-cpp-syntax|jef|1.17.2
azure-pipelines|ms-|1.208.0
vscode-docker|ms-|1.23.3
csharp|ms-|1.25.4
vscode-kubernetes-tools|ms-|1.3.11
remote-containers|ms-|0.278.0
remote-ssh|ms-|0.97.2023020215
remote-ssh-edit|ms-|0.84.0
azure-account|ms-|0.11.3
cmake-tools|ms-|1.13.45
cpptools|ms-|1.14.1
cpptools-extension-pack|ms-|1.3.0
makefile-tools|ms-|0.6.0
powershell-preview|ms-|2023.2.0
remote-explorer|ms-|0.3.2023020609
vsliveshare|ms-|1.0.5828
fabric8-analytics|red|0.3.6
java|red|1.15.2023020803
vscode-xml|red|0.25.2023020103
vscode-yaml|red|1.11.0
sonarlint-vscode|Son|3.14.0
cmake|twx|0.0.17
intellicode-api-usage-examples|Vis|0.2.7
vscodeintellicode|Vis|1.2.30
vscode-boot-dev-pack|vmw|0.2.1
vscode-spring-boot|vmw|1.44.0
vscode-java-debug|vsc|0.48.0
vscode-java-dependency|vsc|0.21.2023013100
vscode-java-pack|vsc|0.25.2022112900
vscode-java-test|vsc|0.37.2023013002
vscode-maven|vsc|0.40.2023020103
vscode-spring-boot-dashboard|vsc|0.10.1
vscode-spring-initializr|vsc|0.11.2

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256198
vstes627cf:30244335
vslsvsres303:30308271
pythonvspyl392:30422396
pythontb:30258533
pythonptprofiler:30281269
vshan820:30294714
pythondataviewer:30285072
vscod805cf:30301675
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593cf:30376535
pythonvs932:30404738
cppdebug:30492333
vsclangdf:30492506
c4g48928:30535728
dsvsc012cf:30540253
pynewext54:30618038
pylantcb52:30590116
pyindex848:30611229
nodejswelcome1:30587009
pyind779:30611226
pythonsymbol12:30651887
6233i204:30659907

```

</details>

### LINKED ISSUE #173839: Include exit code in accessible buffer for failed commands
Sighted users can determine if a command failed via the color of or by hovering the command decoration. That's tedious for a screen reader user, so we should include this info in the accessible buffer.

### LINKED ISSUE #173843: Setting should use markdown description to render properly
Screenshot 2023-02-08 at 11 25 29 AM

### PR DESCRIPTION
**Pull request by meganrogge**
fix #173843
fix #173780
fix #173734
fix [#173839](https://github.com/microsoft/vscode/issues/173839)

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 1]
```diff
@@ -1104,61 +1100,7 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 	}
 
 	focusAccessibilityBuffer(): void {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1102]: standardize on Accessibility -> Accessible?

File: src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts [hunk 2]
```diff
@@ -754,3 +763,76 @@ export class XtermTerminal extends DisposableStore implements IXtermTerminal, II
...
+		@IConfigurationService private readonly _configurationService: IConfigurationService
+	) {
+		super();
+		this.add(this._terminal.registerBufferElementProvider({ provideBufferElements: () => this.focus() }));
+	}
+
+	focus(): DocumentFragment {
+		if (!this._bufferElementFragment) {
+			this._bufferElementFragment = document.createDocumentFragment();
+		}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:786]: This doesn't technically solve https://github.com/microsoft/vscode/issues/173734 but my main goal was to encapsulate things a little more which is accomplished via `AccessibleBuffer`

File: src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts [hunk 3]
```diff
@@ -754,3 +763,76 @@ export class XtermTerminal extends DisposableStore implements IXtermTerminal, II
...
+		for (const command of commands) {
+			header = document.createElement('h2');
+			// without this, the text area gets focused when keyboard shortcuts are used
+			header.tabIndex = -1;
+			header.textContent = command.command.replace(new RegExp(' ', 'g'), '\xA0');
+			if (command.exitCode !== 0) {
+				header.textContent += ` exited with code ${command.exitCode}`;
+			}
+			const output = document.createElement('div');
+			// without this, the text area gets focused when keyboard shortcuts are used
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:817]: text area isn't correct here

**Review by Tyriar — APPROVED** [review]

File: src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:817]: text area of xterm.js is what i'm referring to here

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 1 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1102]: https://github.com/microsoft/vscode/pull/173863

File: src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts (hunk 2 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:786]: 🤔 but we're updating the elements in the callback https://github.com/microsoft/vscode/blob/062ecb0e9657cc5d441135ea5701f9059f7099f2/src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts#L780-L785

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts:786]: You might need to do this to get the screen reader to read it correctly, but this is still a little weird as this means the command triggers `focusAccessibilityBuffer`, which calls into `AccessibleBuffer.focus` and generates the fragment, then xterm.js requests buffer elements and then calls `AccessibleBuffer.focus` again. So this is a little dangerous as it's recursive, I think the only reason it's working fine is calling focus on a focused element won't trigger the `focus` event listener.

I was imagining:

- User focuses the element (either via our command or manually somehow)
- xterm.js listens to focus on the element which calls `provideBufferElements`, this does not focus, it only provides the elements

This is simpler with no recursion.

**Pull request merged**